onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -856,9 +857,22 @@ class PlannerImpl {
                     // we have seen
                     plan_.SetLocation(static_cast<size_t>(index), exec_provider->GetOrtDeviceByMemType(OrtMemType::OrtMemTypeDefault));
                   } else {
-                    // Default the location to CPU
-                    plan_.SetLocation(static_cast<size_t>(index),
-                                      execution_providers_.Get(CPU)->GetOrtDeviceByMemType(OrtMemType::OrtMemTypeDefault));
+                    // We want to minimize the amount of copies, so we want at least one
+                    // device to match or match both if they are CPU based.
+                    OrtDevice result;
+                    auto first_device = already_seen_ep_for_node_arg->second->GetOrtDeviceByMemType(OrtMemType::OrtMemTypeDefault);
+                    auto second_device = exec_provider->GetOrtDeviceByMemType(OrtMemType::OrtMemTypeDefault);
+                    if (first_device.Type() == OrtDevice::CPU && second_device.Type() == OrtDevice::CPU) {
+                      if (first_device.MemType() == OrtDevice::MemType::DEFAULT &&
+                          second_device.MemType() == OrtDevice::MemType::DEFAULT) {
+                        result = (first_device.GetAlignment() >= second_device.GetAlignment()) ? first_device : second_device;
+                      } else {
+                        result = (first_device.MemType() != OrtDevice::MemType::DEFAULT) ? first_device : second_device;
+                      }
+                    } else {
+                      result = (first_device.Type() == OrtDevice::CPU) ? first_device : second_device;
+                    }
```

Comment:
Can we extract the logic comparing two OrtDevice instances from here and below out and use the helpful comments from below in the common code?